strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactGroupServiceImpl.java

```diff
@@ -0,0 +1,92 @@
+package org.carlspring.strongbox.services.impl;
+
+import org.carlspring.strongbox.data.service.CommonCrudService;
+import org.carlspring.strongbox.domain.ArtifactGroup;
+import org.carlspring.strongbox.services.ArtifactGroupService;
+import org.carlspring.strongbox.util.BeanUtils;
+
+import java.lang.reflect.UndeclaredThrowableException;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import com.orientechnologies.common.concur.ONeedRetryException;
+import com.orientechnologies.orient.core.record.impl.ODocument;
+import com.orientechnologies.orient.core.sql.query.OSQLSynchQuery;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * @author Przemyslaw Fusik
+ */
+@Service
+@Transactional
+public class ArtifactGroupServiceImpl
+        extends CommonCrudService<ArtifactGroup>
+        implements ArtifactGroupService
+{
+
+    @Override
+    public Class<ArtifactGroup> getEntityClass()
+    {
+        return ArtifactGroup.class;
+    }
+
+    @Override
+    public <T extends ArtifactGroup> T findOneOrCreate(Class<T> type,
+                                                       Map<String, ? extends Object> properties)
+    {
+        Optional<T> optional = tryFind(type, properties);
+        if (optional.isPresent())
+        {
+            return optional.get();
+        }
+
+        T artifactGroup = instantiateNew(type, properties);
+
+        try
+        {
+            return save(artifactGroup);
+        }
+        catch (ONeedRetryException ex)
+        {
+            optional = tryFind(type, properties);
+            if (optional.isPresent())
+            {
+                return optional.get();
+            }
+            throw ex;
+        }
+    }
+
+    private <T extends ArtifactGroup> T instantiateNew(Class<T> type,
```

Comment:
@fuss86 sorry but I don't really like this way to create `ArtifactGorup` instance.
It's seems to me that we should better have concert `ArtifactGrouptService` implementations, so inherited services can create instances as they need to do it. 
Yeah, you told that you don't want to do it but I don't see other way to provide proper generic implementations.

Reply:
I still would like to avoid this unless it is necessary. We can do it if needed on later time.

Reply:
Sorry, but it's seems that we should have it now.

Reply:
I'm against of doing it, because current implementation may be used in generic way. If there are any problems with it in the future, we will change it. Right now it perfectly suits our needs any I believe it may fit to any others.

Reply:
Guys, what seems to be the problem here?
@sbespalov : How would you like it to look?

Reply:
Done in https://github.com/strongbox/strongbox/pull/987/commits/2dc51fe56eefda5c5822ec9940c0f9f8007b8ec5